docs/CIP_6_Pipelines_Registry.md

```diff
@@ -0,0 +1,309 @@
+
+It is important to preserve the existing behavior of our API where users can simply pass text to the add step and have it be embedded. To do this, we will create a default pipeline that is used when no pipeline is provided. This default pipeline will simply embed the text. This default pipeline will be persisted with the collection, and will be used when no pipeline is provided. This means that users can create a collection with a default pipeline, and then query the collection without having to provide the pipeline. This is a significant improvement over the current system, where users must remember which embedding function they used when they created the collection.
+
+Pipelines can be persisted by serializing the pipeline steps and storing them in the metadata of the collection. This means that when a collection is loaded, the pipeline steps can be deserialized and used to recreate the pipeline. This is a significant improvement over the current system, where users must remember which embedding function they used when they created the collection.
+
+For example, the default pipeline would simply Embed the text, and would be persisted as follows:
+
+```python
+collection = client.create_collection("my_collection", pipeline=DefaultEmbeddingFunctionPipelineStep())
+
+# ... some time later
+
+collection = client.get_collection("my_collection") # The default pipeline is automatically loaded
+```
+
+#### Query Pipelines
+
+In order to support embedding the data automatically for the user on query, we propose the introduction of the query_pipeline, which is the default pipeline run on the query data. This pipeline is persisted with the collection, and is used when no pipeline is provided. This means that users can create a collection with a default query pipeline, and then query the collection without having to provide the query pipeline.
+
+For example, the default query pipeline would simply Embed the text, and would be defined as follows:
+
+```python
+collection = client.create_collection("my_collection", pipeline=DefaultEmbeddingFunctionPipelineStep(), query_pipeline=DefaultEmbeddingFunctionPipelineStep()) # The defaults would not actually have to be specified, as they would be the default values. Here we are specifying them for clarity.
+collection.add(ids, data=["text_to_embed"]) # The content of the data is embedding using the pipeline and then added to the collection
+collection.query(data=["text_to_query"]) # The content of the data is embedding using the query pipeline and then queried against the collection
+```
+
+This allows for the pipelines at insert time, and the pipelines at query time to be different. This is useful for a variety of reasons. For example, if you wanted to chunk the data at insert time, but not at query time, you could do something like this:
+
+```python
+collection = client.create_collection("my_collection", pipeline=Embed(Chunk()), query_pipeline=Embed())
+```
+
+#### The I/0 of a pipeline stage
```

Comment:
@HammadB, do you think it makes sense for pipelines to be allowed to operate over metadata and Ids?

This can be useful for the following:

- Generate Ids
- When doing chunking add metadata associated with chunks

Reply:
It can! You can imagine an AssignIds() step, and I gave a chunking example - did you want something else?

Reply:
Following on the metadata, would this allow for me to change the metadata before inserting alongside the data? For example maybe I have a url and want to get the domain from it and add it as a new metadata field before inserting. Or would that be something to be done before the whole pipeline chaining?

Reply:
I understand that pipeline steps will have access to read and update all - docs, embeddings, ids, metadata.